### PR TITLE
Multiprocessed camera

### PR DESCRIPTION
**Pull request by tlpss**
Goal of this PR is to have a way to consume the camera data streams in multiple processes. 
This is a.o. needed for the cloth competition to visualize the rgb stream in once process and use the stream in another process to control the robot. Need other processes to not take up CPU resources on the control process for visualisations.

Tried a number of things:
- ~Python MP Queues~: not suited, only 'pop' from queue so can be consumed only once
- CycloneDDS: works but cannot deliver 1080p image at more than 1-2FPS over a UPD socket. Limited bandwidth of DDS is known issue in ROS2 community, solution is to resort to shared memory communication, though I think CycloneDDS should be able to do more.
- **Python MP Shared Memory**: can deliver the speed, but need to provide synchronisation manually to avoid simultaneous access and to make read/writes atomic.

with stuff like this, once starts to appreciate everything ROS does under the hood (using DDS). Though throughput is an unsolved issue in ROS 2 as well, see 
- https://discourse.ros.org/t/ros2-speed/20162/23
- https://discourse.ros.org/t/latency-and-throughput-in-ros2/4367

TODOs:
- [x] Document the `camera.multiprocessing` module and also link it in top-level docs of camera-toolkit package
- [ ] protect SHM blocks with Locks
- [x] Make Depth optional, i.e. make it possible to only stream RGB, though at that point it is actually an RGB camera so we maybe should create similar code for the RGB interface. -> RGB & RGBD interface
- [x] change RGB interface with flag that allows to return image in int instead of float (4x reduction in data size), and use that in the publisher/subscriber #61 (separate PR!)
- [ ] write testing with a dummy camera implementation
- [x] provide code to measure FPS/latency of SHM solution -> use ReRun timings as proxy or RGB timestamp
- ~[ ] (@tlpss, optional - dive into DDS to check if we can improve performance (DDS features > hacking SHM solutions ourselves: access control, buffers, QoS,...)~ ignore DDS for now

### CODE REVIEW

**Comment by tlpss** [comment]
@Victorlouisdg, have a go at adding the Locks if you want!

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgbd_camera.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    def _close_shared_memory(self):
+        """Closing shared memory signal that"""
+        self.depth_shm.close()
+        self.depth_image_shm.close()
+
+    def stop_receiving(self):
+        super().stop_receiving()
+        self._close_shared_memory()
+
+
```
[the remark this answers is not in the document]
Comment (Victorlouisdg) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgbd_camera.py:166]: `numpy.rot()` would indeed be better, apparently, it takes a `k` parameter to specify the number of times you want to rotate 90 degrees. Do you think we should expose that?

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        dummy_intrinsics = self._camera.intrinsics_matrix()
+
+        self.rgb_shm = shared_memory.SharedMemory(
+            create=True,
+            size=dummy_rgb_image.nbytes,
+            name=f"{self._shared_memory_namespace}_{_RGB_SHM_NAME}",
+        )
+        self.rgb_shm_array = np.ndarray(dummy_rgb_image.shape, dtype=dummy_rgb_image.dtype, buffer=self.rgb_shm.buf)
+
+        self.rgb_timestamp_shm = shared_memory.SharedMemory(
```
[the remark this answers is not in the document]
Comment (Victorlouisdg) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py:62]: True, but then you need to explicitly pass this `Array` to all receivers right?

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        self.intrinsics_shm.close()
+
+    def stop_receiving(self) -> None:
+        self._close_shared_memory()
+
+    def __del__(self) -> None:
+        self._close_shared_memory()
+
+
+class MultiProcessRGBLogger(Process):
```
[the remark this answers is not in the document]
Comment (Victorlouisdg) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py:178]: Agreed

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+        previous_timestamp = time.time()
+
+        while not self.shutdown_event.is_set():
+            timestamp = self.multiProcessRGBReceiver.get_rgb_image_timestamp()
+            if timestamp > previous_timestamp:
+                image = self.multiProcessRGBReceiver.get_rgb_image()
+                rerun.log_image(self._shared_memory_namespace, image)
+                previous_timestamp = timestamp
+            time.sleep(0.001)  # Check every millisecond
```
[the remark this answers is not in the document]
Comment (Victorlouisdg) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py:210]: The publisher and receiver are not in sync, so increasing this delay reduces the responsiveness of the receiver. Now if the publisher writes a new timestamp, the receiver will likely pick this up within a millisecond. In my opinion, the possible savings in CPU cycles is not worth the additional latency.

[the remark this answers is not in the document]
Comment (tlpss) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py:210]: > the possible savings in CPU cycles is not worth the additional latency.
agreed

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tlpss) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py:62]: ah yeah right. As long as we don't explicitly use the Lock primitives we don't have to pass anything.. So then I agree it makes sense to use SHM, even though we should probably explicitly document the possible synchronization issues

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgbd_camera.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tlpss) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgbd_camera.py:166]: I would expose the param in degrees, then internally validate it is a multiple of 90 and convert it to the `k` param. It's more work for the developer, but it is the most natural interface for the user I guess. Not sure if we should spend time on it though, it's a detail

[the remark this answers is not in the document]
Comment (Victorlouisdg) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgbd_camera.py:166]: Done

**Comment by tlpss** [comment]
@Victorlouisdg , what is needed before we can merge this?

**Comment by Victorlouisdg** [comment]
I would like to land PR #64 first and utilize the `ImageTransforms` in the `MultiProcessRGBLogger` class.

File: airo-camera-toolkit/airo_camera_toolkit/cameras/zed2i.py
[the remark this answers is not in the document]
Comment (tlpss) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/zed2i.py:1]: For me you can add this here, as long as we add it in the PR description then

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_video_recorder.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            timestamp = datetime.datetime.now().strftime("%H:%M:%S:%f")[:-3]
+            video_name = f"{timestamp}.mp4"
+            video_path = os.path.join(output_dir, video_name)
+            video_path = os.path.abspath(video_path)
+
+        self._video_path = video_path
+
+    def run(self) -> None:
+        """main loop of the process, runs until the process is terminated"""
+        import ffmpegcv  # type: ignore
```
Comment (Victorlouisdg) [inline airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_video_recorder.py:79]: Document that this python package [ffmpegcv](https://github.com/chenxinfeng4/ffmpegcv) and [ffmpeg 6.0](https://anaconda.org/conda-forge/ffmpeg) (available via conda) must be installed for (hvec) video recording.

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_video_recorder.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+import loguru
+from airo_camera_toolkit.cameras.multiprocess.multiprocess_rgb_camera import MultiprocessRGBReceiver
+from airo_camera_toolkit.image_transforms.image_transform import ImageTransform
+from airo_camera_toolkit.utils import ImageConverter
+
+logger = loguru.logger
+
+
+class FPSMonitor:
```
[the remark this answers is not in the document]
Comment (Victorlouisdg) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_video_recorder.py]: This class has been removed from this PR.

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_video_recorder.py (hunk 5 above)
Comment (Victorlouisdg) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_video_recorder.py:37]: Done.

File: airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+        self.rgb_shm_array: np.ndarray = np.ndarray(
+            (camera_resolution_height, camera_resolution_width, 3),
+            dtype=np.float32,
+            buffer=self.rgb_shm.buf,
+        )
+
+        self.intrinsics_shm_array: np.ndarray = np.ndarray((3, 3), dtype=np.float64, buffer=self.intrinsics_shm.buf)
+
+    def get_rgb_image_timestamp(self) -> float:
+        return np.frombuffer(self.rgb_timestamp_shm.buf, dtype=np.float64)[0]
```
[the remark this answers is not in the document]
Comment (Victorlouisdg) [reply airo-camera-toolkit/airo_camera_toolkit/cameras/multiprocess/multiprocess_rgb_camera.py]: I added a note to warn for this.

**Pull request merged**